<h1 style="color:#CD7F32;">BRONZE LAYER</h1>


In [10]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
# Evaluating data processing
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [7]:
warnings.filterwarnings('ignore')

In [4]:
df = pd.read_csv('US_Accidents_March23.csv')

In [5]:
df.head(10)

ID   Source  Severity           Start_Time             End_Time  \
0   A-1  Source2         3  2016-02-08 05:46:00  2016-02-08 11:00:00   
1   A-2  Source2         2  2016-02-08 06:07:59  2016-02-08 06:37:59   
2   A-3  Source2         2  2016-02-08 06:49:27  2016-02-08 07:19:27   
3   A-4  Source2         3  2016-02-08 07:23:34  2016-02-08 07:53:34   
4   A-5  Source2         2  2016-02-08 07:39:07  2016-02-08 08:09:07   
5   A-6  Source2         3  2016-02-08 07:44:26  2016-02-08 08:14:26   
6   A-7  Source2         2  2016-02-08 07:59:35  2016-02-08 08:29:35   
7   A-8  Source2         3  2016-02-08 07:59:58  2016-02-08 08:29:58   
8   A-9  Source2         2  2016-02-08 08:00:40  2016-02-08 08:30:40   
9  A-10  Source2         3  2016-02-08 08:10:04  2016-02-08 08:40:04   

   Start_Lat  Start_Lng  End_Lat  End_Lng  Distance(mi)  ... Roundabout  \
0  39.865147 -84.058723      NaN      NaN          0.01  ...      False   
1  39.928059 -82.831184      NaN      NaN          0.01  ...      False   
2  39.063148 -84.032608      NaN      NaN          0.01  ...      False   
3  39.747753 -84.205582      NaN      NaN          0.01  ...      False   
4  39.627781 -84.188354      NaN      NaN          0.01  ...      False   
5  40.100590 -82.925194      NaN      NaN          0.01  ...      False   
6  39.758274 -84.230507      NaN      NaN          0.00  ...      False   
7  39.770382 -84.194901      NaN      NaN          0.01  ...      False   
8  39.778061 -84.172005      NaN      NaN          0.00  ...      False   
9  40.100590 -82.925194      NaN      NaN          0.01  ...      False   

  Station   Stop Traffic_Calming Traffic_Signal Turning_Loop Sunrise_Sunset  \
0   False  False           False          False        False          Night   
1   False  False           False          False        False          Night   
2   False  False           False           True        False          Night   
3   False  False           False          False        False          Night   
4   False  False           False           True        False            Day   
5   False  False           False          False        False            Day   
6   False  False           False          False        False            Day   
7   False  False           False          False        False            Day   
8   False  False           False          False        False            Day   
9   False  False           False          False        False            Day   

  Civil_Twilight Nautical_Twilight Astronomical_Twilight  
0          Night             Night                 Night  
1          Night             Night                   Day  
2          Night               Day                   Day  
3            Day               Day                   Day  
4            Day               Day                   Day  
5            Day               Day                   Day  
6            Day               Day                   Day  
7            Day               Day                   Day  
8            Day               Day                   Day  
9            Day               Day                   Day  

[10 rows x 46 columns]

In [6]:
#searching about duplicated values
df.duplicated().sum()

0

In [7]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7728394 entries, 0 to 7728393
Data columns (total 46 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   ID                     object 
 1   Source                 object 
 2   Severity               int64  
 3   Start_Time             object 
 4   End_Time               object 
 5   Start_Lat              float64
 6   Start_Lng              float64
 7   End_Lat                float64
 8   End_Lng                float64
 9   Distance(mi)           float64
 10  Description            object 
 11  Street                 object 
 12  City                   object 
 13  County                 object 
 14  State                  object 
 15  Zipcode                object 
 16  Country                object 
 17  Timezone               object 
 18  Airport_Code           object 
 19  Weather_Timestamp      object 
 20  Temperature(F)         float64
 21  Wind_Chill(F)          float64
 22  Humidity(%)       

In [8]:
# Check for null values
df.isnull().sum()

ID                             0
Source                         0
Severity                       0
Start_Time                     0
End_Time                       0
Start_Lat                      0
Start_Lng                      0
End_Lat                  3402762
End_Lng                  3402762
Distance(mi)                   0
Description                    5
Street                     10869
City                         253
County                         0
State                          0
Zipcode                     1915
Country                        0
Timezone                    7808
Airport_Code               22635
Weather_Timestamp         120228
Temperature(F)            163853
Wind_Chill(F)            1999019
Humidity(%)               174144
Pressure(in)              140679
Visibility(mi)            177098
Wind_Direction            175206
Wind_Speed(mph)           571233
Precipitation(in)        2203586
Weather_Condition         173459
Amenity                        0
Bump      

*Filling missing values with interpolate method*

Preserves Data Trends: Forward and backward filling can help maintain the trend and continuity of the data, which is especially useful in time series data.

Handles Consecutive NaNs: The limit parameter allows control over how many consecutive NaNs are filled, preventing excessive filling that might distort the data.

limit is Maximum number of consecutive NaNs to fill. Must be greater than 0.
This method fills missing values by propagating the last valid value forward.
This will forward-fill up to 10 consecutive NaN values with the last valid value.


In [11]:
df.ffill(limit=10, inplace=True)

After forward-filling, this will back-fill up to 5 consecutive NaN values with the next valid value.
This method fills missing values by using the next valid value to fill backward.

In [13]:
df.bfill(limit=10, inplace=True)

Using both ffill and bfill ensures that missing values are filled from both directions—forward and backward. This approach can be useful when you want to ensure data continuity without introducing bias from just one direction.

In [15]:
df.isnull().sum()

ID                             0
Source                         0
Severity                       0
Start_Time                     0
End_Time                       0
Start_Lat                      0
Start_Lng                      0
End_Lat                  3402752
End_Lng                  3402752
Distance(mi)                   0
Description                    0
Street                         0
City                           0
County                         0
State                          0
Zipcode                        0
Country                        0
Timezone                       0
Airport_Code                   0
Weather_Timestamp           3331
Temperature(F)              3333
Wind_Chill(F)            1243204
Humidity(%)                 3333
Pressure(in)                3331
Visibility(mi)              3331
Wind_Direction              3334
Wind_Speed(mph)             4623
Precipitation(in)        1114714
Weather_Condition           3331
Amenity                        0
Bump      

In [16]:
df.shape  

(7728394, 46)

In [18]:
# Drop the columns 'End_Lat' and 'End_Lng' from the DataFrame 'df'
df.drop(columns=['End_Lat', 'End_Lng'], inplace=True)

# Remove rows where any of the specified columns have missing values
df.dropna(subset=['Wind_Chill(F)', 'Wind_Speed(mph)', 'Precipitation(in)'], inplace=True)

In [19]:
df.columns

Index(['ID', 'Source', 'Severity', 'Start_Time', 'End_Time', 'Start_Lat',
       'Start_Lng', 'Distance(mi)', 'Description', 'Street', 'City', 'County',
       'State', 'Zipcode', 'Country', 'Timezone', 'Airport_Code',
       'Weather_Timestamp', 'Temperature(F)', 'Wind_Chill(F)', 'Humidity(%)',
       'Pressure(in)', 'Visibility(mi)', 'Wind_Direction', 'Wind_Speed(mph)',
       'Precipitation(in)', 'Weather_Condition', 'Amenity', 'Bump', 'Crossing',
       'Give_Way', 'Junction', 'No_Exit', 'Railway', 'Roundabout', 'Station',
       'Stop', 'Traffic_Calming', 'Traffic_Signal', 'Turning_Loop',
       'Sunrise_Sunset', 'Civil_Twilight', 'Nautical_Twilight',
       'Astronomical_Twilight'],
      dtype='object')

In [20]:
df.isnull().sum()

ID                       0
Source                   0
Severity                 0
Start_Time               0
End_Time                 0
Start_Lat                0
Start_Lng                0
Distance(mi)             0
Description              0
Street                   0
City                     0
County                   0
State                    0
Zipcode                  0
Country                  0
Timezone                 0
Airport_Code             0
Weather_Timestamp        0
Temperature(F)           0
Wind_Chill(F)            0
Humidity(%)              0
Pressure(in)             0
Visibility(mi)           0
Wind_Direction           0
Wind_Speed(mph)          0
Precipitation(in)        0
Weather_Condition        0
Amenity                  0
Bump                     0
Crossing                 0
Give_Way                 0
Junction                 0
No_Exit                  0
Railway                  0
Roundabout               0
Station                  0
Stop                     0
T

In [21]:
df.describe()

Severity     Start_Lat     Start_Lng  Distance(mi)  Temperature(F)  \
count  6.142287e+06  6.142287e+06  6.142287e+06  6.142287e+06    6.142287e+06   
mean   2.173995e+00  3.625293e+01 -9.484108e+01  6.460911e-01    6.065101e+01   
std    4.682526e-01  5.118693e+00  1.755349e+01  1.861972e+00    1.933470e+01   
min    1.000000e+00  2.455480e+01 -1.245481e+02  0.000000e+00   -8.900000e+01   
25%    2.000000e+00  3.346158e+01 -1.173020e+02  0.000000e+00    4.800000e+01   
50%    2.000000e+00  3.591510e+01 -8.763713e+01  8.300000e-02    6.300000e+01   
75%    2.000000e+00  4.012797e+01 -8.032258e+01  5.960000e-01    7.500000e+01   
max    4.000000e+00  4.900220e+01 -6.748413e+01  4.417500e+02    2.070000e+02   

       Wind_Chill(F)   Humidity(%)  Pressure(in)  Visibility(mi)  \
count   6.142287e+06  6.142287e+06  6.142287e+06    6.142287e+06   
mean    5.868330e+01  6.508806e+01  2.940720e+01    9.005185e+00   
std     2.202547e+01  2.295043e+01  1.089370e+00    2.693157e+00   
min    -8.900000e+01  1.000000e+00  0.000000e+00    0.000000e+00   
25%     4.300000e+01  4.800000e+01  2.922000e+01    1.000000e+01   
50%     6.200000e+01  6.800000e+01  2.975000e+01    1.000000e+01   
75%     7.500000e+01  8.500000e+01  2.998000e+01    1.000000e+01   
max     2.070000e+02  1.000000e+02  5.863000e+01    1.400000e+02   

       Wind_Speed(mph)  Precipitation(in)  
count     6.142287e+06       6.142287e+06  
mean      7.462977e+00       1.062292e-02  
std       5.522065e+00       1.993722e-01  
min       0.000000e+00       0.000000e+00  
25%       3.500000e+00       0.000000e+00  
50%       7.000000e+00       0.000000e+00  
75%       1.000000e+01       0.000000e+00  
max       1.087000e+03       3.647000e+01

Removing outliers
By comparing mean with max & mean with std we got that:
  *  In Start_Lat column, std is greater than mean --> outlier.
  *  In Start_Lng column, std is greater than mean --> outlier.
  *  In Severity column, std is greater than mean --> outlier.
  *  In Pressure(in) column, std is greater than mean --> outlier.

In [24]:
columns = ['Start_Lat', 'Start_Lng'] 

for col in columns:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR #Anything below this boundary is considered an outlier.
    upper_bound = Q3 + 1.5 * IQR #Anything above this boundary is considered an outlier.
    df = df[(df[col] >= lower_bound) & (df[col] <= upper_bound)] 
 #filters the DataFrame df to keep only the rows where
 #the value in the current column lies within the lower_bound and upper_bound

In [63]:
df.to_csv('bronzelayerUSAccidents.csv', index=False)

The target is the feature "severity"
(The level of impact the accident had on the traffic,
1 indicates the least amount of impact.)

In [61]:
# TO MEASURE THE ACCURACY OF DATA AFTER PROCESSING TO PREDICT THE TARGET VARIABLE, THE SEVERITY
# Performing machine learning classification using the RandomForestClassifier on a sampled dataset.
# First, a random sample (10% of the original dataset) is taken and missing values are dropped.
# Categorical columns are then converted into numeric form using LabelEncoder to allow for numerical analysis.
# The features and target are defined, with the Severity column as the target.
# The dataset is split into training and testing sets, using 70% of the data for training and 30% for testing.
# A RandomForestClassifier is initialized and trained on the training data.

sampled_dff = df.sample(frac=0.1, random_state=42).dropna()

categorical_cols = sampled_dff.select_dtypes(include=['object', 'bool']).columns
label_encoder = LabelEncoder()
for col in categorical_cols:
    sampled_dff[col] = label_encoder.fit_transform(sampled_dff[col])
numeric_cols = sampled_dff.select_dtypes(include=['number']).columns
X = sampled_dff[numeric_cols].drop('Severity', axis=1, errors='ignore')
y = sampled_dff['Severity']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)
rf_classifier = RandomForestClassifier(n_jobs=-1, random_state=42) 
rf_classifier.fit(X_train, y_train)
y_pred = rf_classifier.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")

confusion_mat = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(confusion_mat)

class_report = classification_report(y_test, y_pred)
print("Classification Report:")
print(class_report)

Accuracy: 0.93
Confusion Matrix:
[[  1309    622     89      0]
 [   271 149516   3246    221]
 [    27   5368  18694      9]
 [     7   2465    424   2001]]
Classification Report:
              precision    recall  f1-score   support

           1       0.81      0.65      0.72      2020
           2       0.95      0.98      0.96    153254
           3       0.83      0.78      0.80     24098
           4       0.90      0.41      0.56      4897

    accuracy                           0.93    184269
   macro avg       0.87      0.70      0.76    184269
weighted avg       0.93      0.93      0.93    184269



Class 1: Minor severity
Class 2: Moderate severity
Class 3: Major severity
Class 4: Critical severity

The confusion matrix shows the performance of the classification model by comparing the actual and predicted classifications. Each row represents the actual class, while each column represents the predicted class.
Class 1 (Severity 1):
Correctly classified: 1309,
Misclassified as Class 2: 622,
Misclassified as Class 3: 89,
Misclassified as Class 4: 0
Class 2 (Severity 2):
Correctly classified: 149516,
Misclassified as Class 1: 271,
Misclassified as Class 3: 3246,
Misclassified as Class 4: 221
Class 3 (Severity 3):
Correctly classified: 18694,
Misclassified as Class 1: 27,
Misclassified as Class 2: 5368,
Misclassified as Class 4: 9
Class 4 (Severity 4):
Correctly classified: 2001,
Misclassified as Class 1: 7,
Misclassified as Class 2: 2465,
Misclassified as Class 3: 424